### DISCUSSION TITLE
RM3 WAMIT example: strange peaks

### DISCUSSION THREAD

**Discussion by antbog** [WEC-Sim/WEC-Sim#1244] [Q&A]
**Note: italicized text below is include as an example and should be updated before submission. If you feel any section is not applicable to your request, please replace with 'N/A' rather than delete a section.**

**Describe the bug**
I launched a WAMIT run for the RM3 WEC provided in the tutorials. I used the input files given in this GitHub repository (examples > BEMIO > WAMIT > RM3) and changed only the IOPTN(4) from 0 to 2 in the .frc file.

When I display the frequency-dependent quantities (using bemio.m or my own routines), it appears that strange peaks that weren't in the tutorial's .out appeared. They could be due to irregular frequencies (but the option IRR=1 still stands in .cfg) or resonances. Anyway, I don't understand the emergence of these peaks and especially why you don't have it in the tutorial hydroData figures.

Moreover, even when I don't change the IOPTN(4) (ie I keep all the exact settings provided in the examples > BEMIO > WAMIT > RM3 files), the peaks still appear.

**WEC-Sim file(s)**
examples > BEMIO > WAMIT > RM3 all input files (*.wam, *.gdf, *.dat, , *.cfg, *.frc, *.pot)

**To Reproduce**
1. Launch WAMIT with these files as input
2. Use the .out file as input for bemio.m or for any routine allowing to visualize the frequency-dependent quantities, or even the response functions (it this case, we observe beat instead of peaks).

**Expected behavior**
The peaks shouldn't be there.

**Screenshots**
Plots from the RM3 tutorial:
![Added_Mass](https://github.com/WEC-Sim/WEC-Sim/assets/135218472/8bacd89c-05f1-4162-889d-20d54a877743)

Plot that I obtain after WAMIT calculations with the input files provided in examples > BEMIO > WAMIT > RM3
![Added_Mass_irr](https://github.com/WEC-Sim/WEC-Sim/assets/135218472/12d454bf-ee9a-4a66-83ae-ac80cc068af7)

**Workstation Information (this section MUST BE completed):**
OS: Windows 10 Enterprise, Version 22H2
[MATLAB/Simulink Release](https://www.mathworks.com/help/matlab/release-notes.html) 9.14.0.2254940 R2023a
[WEC-Sim Release](http://wec-sim.github.io/WEC-Sim/man/release_notes.html) v5.0.1
WAMIT Version 7.4(x64)

**Additional context**
N/A

**Comment by salhus** [comment]
Hi @antbog ,

This is a known nuance of Boundary Element Methods that WAMIT and other software use. This happens when the mesh has an internal tank and is referred to as an 'irregular frequency' in literature. WAMIT has an irregular frequency removal feature which works for most applications.
![image](https://github.com/WEC-Sim/WEC-Sim/assets/84348506/4cc9f9c1-c876-4027-9666-f498897dfedf)

 This can be done by setting IRR = 1 or 2 in the WAMIT runs.
![image](https://github.com/WEC-Sim/WEC-Sim/assets/84348506/f3b218a2-89df-4c40-b07c-9507ce41bdd7)

Additionally, research has been done on removing such discontinuities by post-processing the hydrodynamics coefficients. Here is an example,

https://asmedigitalcollection.asme.org/OMAE/proceedings/OMAE2019/58882/V009T12A022/1068046

Cheers,
Sal Husain, PhD,
National Renewable Energy Laboratory.

**Comment by dforbush2** [comment]
Hi @salhus and @antbog 

It looks like @antbog still does have IRR=1 in his input file since he is using our input files in examples/BEMIO/WAMIT/RM3. The question is why do these peaks emerge here, but are NOT present in the tutorial BEM result which allegedly come from the same results. Changing IOPTN(4) should not affect any thing here, its just toggling an RAO output on/off. 

In this directory IRR =1, so I would expect @antbog to be getting the same results as in the tutorial, especially as he is using the same WAMIT version we are.

**Comment by antbog** [comment]
Hi @salhus and @dforbush2 

Yes, I'm aware of the existence of this option removing the effects of irregular frequencies and as @dforbush2 said, I'm using your input files so in particular, IRR=1 is in the .cfg and the .gdf contains correctly the interior free surface panels. 

Here is the .out I got if you want to check. 
[rm3out.txt](https://github.com/WEC-Sim/WEC-Sim/files/12426107/rm3out.txt)

I superimposed my curves with the ones from the tutorial .out and it shows that apart from the peaks, they are similar.

I wonder if it has something to do with the contact between the float hole panels surrounding some of the spar panels:
![RM3meshing](https://github.com/WEC-Sim/WEC-Sim/assets/135218472/89107402-d0fb-4c19-b0e0-1605578e71db)

But in any case I don't understand why and don't get the same results as in the tutorial.

Cheers,
Antoine Bogenschutz

**Comment by dav-og** [comment]
Yes, the agreement does look good apart from the irregular frequencies:

![image](https://github.com/WEC-Sim/WEC-Sim/assets/12831266/6d48c180-5e44-459a-a3cd-5f6bed1cf492)

For some reason the irregular frequency removal option is not working properly in your case. However, your .out file shows the option is on. Hence, it could be something related to the meshes (e.g. maybe the mesh of the interior free surface is not being created properly).

In any case, there must be a discrepancy somewhere in your input files (and/or meshes) vs the input files in `examples/BEMIO/WAMIT/RM3`. I've tried to compare the two in BEMRosetta, but nothing jumps out:

![image](https://github.com/WEC-Sim/WEC-Sim/assets/12831266/039ea5c6-1dab-4d1d-a01b-cc2335bb5f9e)

(same CB, CG, Vsub, etc)

Without access to the rest of your input files and meshes its impossible to say.

Cheers,
Dave

**Comment by antbog** [comment]
Hi, 

Here are all the input files and meshes. 

[0823_inutFiles.zip](https://github.com/WEC-Sim/WEC-Sim/files/12428364/0823_inutFiles.zip)

They are supposed to be the same than those on the Git repository as I copy-pasted them from there.

Cheers,
Antoine

**Comment by dav-og** [comment]
Its a strange one - nothing jumps out to me...

However, I can see that the files you shared are not _exactly_ the same as in the examples folder. I've noticed some slight differences in the .pot and .cfg files for example:

![image](https://github.com/WEC-Sim/WEC-Sim/assets/12831266/282f280d-0451-4c73-b137-00098aa74f60)

![image](https://github.com/WEC-Sim/WEC-Sim/assets/12831266/3d58e37b-2d17-4492-97b3-6a3bfe170b93)

So it looks like the input files have been modified. I don't see how any of these changes would introduce irregular frequencies in your results...but there could be another difference that I haven't caught.

Hence, I'd suggest just starting again ensuring that you have the _exact_ same files from the examples folder...run WAMIT with those files and see if you still get irregular frequencies.

Cheers,
Dave

**Comment by antbog** [comment]
Okay so did you just run WAMIT with my input files without any peaks appearing? That's so strange, but then it might be related to my computer I don't know ...

Yes absolutely I removed the computation of the infinite period because I have a further program that doesn't support it as input and I set the RAMGBMAX in config.wam instead of here. I try it rn.

Cheers,
Antoine

**Comment by dav-og** [comment]
I didn't re-run WAMIT - I used some existing outputs. Maybe @salhus or @dforbush2 can re-run the input files from `examples/BEMIO/WAMIT/RM3` and verify that the results still have no irregular frequencies.

**Comment by salhus** [comment]
@antbog 
Hi Antoine,
Apologies for not reading your issue description carefully. Let me get back to you soon. As others have pointed, this does seem to be an unexpected issue.
Cheers,
Sal

**Comment by dforbush2** [comment]
I'm running this case as we speak and will follow up as soon as it finishes. A few remarks:
1) @dav-og good eye. As far as I can see the differences here shouldn't make any difference w.r.t the calculation method, the change in the PER vector just removes the infinite frequency calculation.
2 @antbog though its counter-intuitive if you are used to the  conventions of other BEM solvers, the WAMIT convention (for use with WEC-Sim at least) is that we model each body with its local origin at the CG and then specify its global position in the *pot file. So your plot of the superimposed gdf that shows interference is not actually what should be seen by WAMIT. 

Is it possible this is a version issue? I see version 7.2 in our example output file... and note you're running v7.4.

**Comment by dav-og** [comment]
I wondered the same (v7.2 vs v7.4) but nothing obvious from the release notes:

![image](https://github.com/WEC-Sim/WEC-Sim/assets/12831266/c3b06ced-5600-4d3f-89c7-4731067a92ba)

![image](https://github.com/WEC-Sim/WEC-Sim/assets/12831266/fde916ff-5f56-4b77-a4cd-e9fe59305397)

https://www.wamit.com/manualv7.4/wamit_v74manualch1.html#x2-70001.6

...unless its the "new subroutine" 🤔

**Comment by dforbush2** [comment]
Checking the repo, it is the EXACT same WAMIT run that lives in examples/RM3, examples/BEMO/WAMIT/RM3/ and the RM3 tutorial (you can tell by the time stamps in the *out file). And running BEMIO on all, they are the exact same. These indicate that they are v7.2 outputs and they DO NOT have these spikes. 

Comparing to my fresh v7.4 WAMIT run of the inputs provided by @antbog to those found in  examples/BEMO/WAMIT/RM3/,  the outputs are the same and the spikes DO appear.

It is _possible_ that the *out file we have in the repo was long ago manually cleaned up (by a developer whose identity has faded into legend) to remove those spikes: my finishing v7.2 run will confirm this. If is shows the spikes, then this is what happened.  

**If if does NOT have spikes, then this is a WAMIT version issue that surfaces in v7.4. Developers should be notified. This has some serious implications and we should get to the bottom of it.**

@antbog Note that the internal "hole" of the float will cause some singularities that won't be corrected by the IRR. Assuming deep water, the radian wave frequency corresponding to the diameter of this hole is ~3.2 rad/s, which is about where the mess is. 

@salhus I do note that in your comparison the off-diagonal K terms are significantly different between the two runs, but they are the same in both of my outputs. That's odd and worth investigating.  

I still have a v7.2 run finishing, and will confirm when it finishes (I had to run this locally).

**Comment by salhus** [comment]
Hi @dforbush2,
I think the above comparison is between @antbog 's out file and wecsim's existing rm3 out file. 
I ran a new run using the mesh files provided by @antbog (which is I think identical), and this is what I got (I wrote my own input files and didnt use wec-sim's rm3 input files) 
![image](https://github.com/WEC-Sim/WEC-Sim/assets/84348506/3b0799b5-c5e8-4ff0-a27b-c4de8cbc54dd)


I am about to finish another run where I set the IRR=2, will update soon.

**Comment by salhus** [comment]
Also, I don't think IOPTN(4) is relevant to the spikes. Here is why.
I used the infinite frequency option and still got the spikes, and looking at the IRF based infinite added mass calculations the spikes do show up 
![image](https://github.com/WEC-Sim/WEC-Sim/assets/84348506/ec4f5fa2-daea-4c52-91e2-dcee1ae69f7b)

However, the inf. added mass calculated by enabling IOPTN(4) gives similar values,
![image](https://github.com/WEC-Sim/WEC-Sim/assets/84348506/65ebb3a1-03ed-4cf5-97e2-fd7e4286c563)

**Comment by dforbush2** [comment]
We are in correspondence with WAMIT w.r.t this issue. While we do not yet have an understanding of why v7.2 differs from v7.4/v7.4.1, we have learned that IRR > 0 may not work reliably for bodies with contacting surfaces, as we have in this case near the free surface between the float and spar (as @antbog pointed out). 

Note that IRR =1 DOES appear to work for this case in v7.2: we're still working on understanding why this is, and, at least for the RM3 geometry, IRR = 1 in v7.4 does not appear to be working correctly even for a float only run. 

EDIT @salhus (below) is right. The block diagonal terms are sensitive to other bodies in that the boundary conditions of these terms differ when other physical bodies are present nearby and this will affect their solution.

**Comment by salhus** [comment]
@dforbush2 
Running them separately doesn't make sense as the hydrodynamic couplings will be absent. 
I think running them separately was a way to check if the IRR is working not a prescription for running such a case.
I think like I said in a separate conversation at OREC,  I would recommend introducing a small gap between the float and the spar. As that way there won't be any intersecting normals. 

@antbog 
Either way, we would update the example case soon. This thread will be updated.

Cheers,
Sal

**Comment by dforbush2** [comment]
We have gained more insight into the results of 7.4 and 7.4.1. 
![Added_Mass](https://github.com/WEC-Sim/WEC-Sim/assets/30122876/c460240c-1228-49cc-956a-01e9d44be12e)

These remaining frequencies are irregular frequencies that are not removed due to the contacting surfaces (exterior of spar cylinder, interior of float moonpool) causing the problem to be ill-conditioned. IRR = 1 "worked" in 7.2, but, according to WAMIT developers, this was not a robust result and this meshing/modeling approach is no longer best practice in v7.4 and v7.4.1. 

A study was performed in which the spar diameter was reduced. The results show that these irregular frequencies are removed, but a low frequency "pumping mode" resonance due to the float moonpool is NOT removed via IRR. It is "physical" in the sense that it is a real free surface, but is not realistic because in reality it will be sharply attenuated by viscosity/vortex formation etc. This feature is qualitatively robust to different spar diameter selections.
![Added_Mass](https://github.com/WEC-Sim/WEC-Sim/assets/30122876/531e72dc-2b4d-4974-9eab-5be0bfdb49b6)

This implies that a "correct" approach will 1) not have interfering panels such that IRRs can be removed  and 2) represent the float in such a way that the pumping (and sloshing modes in surge and pitch, equally pronounced and problematic) are not present.

An attempt was performed with a shortened spar, such that there were no contacting panels between the float and spar. This resulted in the following: which does not appear improved.
![Added_Mass](https://github.com/WEC-Sim/WEC-Sim/assets/30122876/1022b706-abe8-4816-b62b-66b7a2bd21b6)

Work continues and will inform future best practices. In the mean time @antbog see PR #1076 for a function that may be useful to you.

**Comment by antbog** [comment]
Hi @hachikoi1, 
I didn't create the circular mesh on Rhino, I only opened the .gdf files provided by WEC-Sim (examples > BEMIO > WAMIT > RM3)  on this software. So unfortunately I don't have answer to your question.

@dav-og @dforbush2 @salhus thank you for these updates, I'm looking forward for future ones and will look at #1076 function.

Cheers,
Antoine

**Comment by salhus** [comment]
Hi Antoine @antbog,

Thanks for brining this issue. We are investigating this and developing a better example. Sorry about the long development time. Hopefully, this should be resolved in the next few weeks. This thread will be updated when the development efforts are complete.

Cheers,
Sal

**Comment by TianyuanWangi** [comment]
Hi [salhus](https://github.com/salhus)
Here I share some manuscripts regarding the strange peaks in the excitation loads and radiaton coefficients
[A viscous damping model for piston mode resona.pdf](https://github.com/WEC-Sim/WEC-Sim/files/12755989/A.viscous.damping.model.for.piston.mode.resona.pdf)
[Relative Heave Motion Responses of Floating Du.pdf](https://github.com/WEC-Sim/WEC-Sim/files/12755990/Relative.Heave.Motion.Responses.of.Floating.Du.pdf)
[Hydrodynamic coefficients in heave of a moonpo.pdf](https://github.com/WEC-Sim/WEC-Sim/files/12755992/Hydrodynamic.coefficients.in.heave.of.a.moonpo.pdf)
From these three papers, the strange peaks are caused by the narrow gap between two bodies and they do exist in physics. It seems to be related to the trapping phenomenoon in the moonpool, similar to the phenomenoon mentioned in your thesis. But the potential flow solution overestimates the wave elevation in the gap. The radiation coefficients calculated based on CFD also have strange peaks, see the third paper.
A general solution to this issue is to add a lid (dampling) to the gap (on the water surface) to avoid overestimating the water elevation in the gap. But the damping coefficient is based on the experimental or CFD results.
From my point, the overestimation of radiation wave would obtain incorrect added mass and radiation damping coefficients around the resonant frequency. In time-domain calculations, turning off convolution does not affect results away from the resonant frequency, because WEC-Sim obtains added mass and radiation damping coefficients by interpolation. However, if the convolution is turned on, incorrect added mass and radiation damping coefficients will affect the results over the entire frequency range. It is therefore necessary to obtain correct hydrodynamic coefficients before performing time-domain calculations
I hope this helps, and I'm looking forward to your deveploment.
Best regards,
Tianyuan

**Comment by salhus** [comment]
Hi Tianyuan @hachikoi1 ,

Thanks for sharing those papers. And it is rather fascinating that the peaks are in fact physical. We will try to finalize if any changes are needed for this example, and update this thread in the coming weeks.

Cheers,
Sal

**Comment by TianyuanWangi** [comment]
Hi 

Any progress on this issue? It's a question that also bothers me. 

From what I understand, in the linear frequency-domain calculations, the wave-excitation loads are obtained by fixing the two objects without any motion. And the wave-radiation loads are obtained by forcing one of the objects to move with unit ampitude in one direction.

For example, added mass matrix of the two objects has a size of 12*12, the added mass A(1,1) is obtained by forcing object 1 to move with unit ampitude in the surge direction, and A(1,7) is the added mass imposed on object 2 by the forced motion of object 1 in the surge direction. Likewise, A(7,7) is obtained by forcing object 2 to move in the surge direction, and A(7,1) is the added mass imposed on object 1 by the forced motion of object 2 in the surge direction.

Considering the existence of rigid constraints between the two bodies, should the radiation coefficients be calculated in the surge, sway, roll and pitch directions? For the outer bodies with a moonpool, the forced motions of one of the two objects in the four directions would cause large-amplitude radiation waves, especially if the gap between the two objects is small.

I am now worrying whether the strange peaks in the surge, sway, roll and pitch directions are caused by the narrow gap resonance? Or generated by the forced motions? If the strange peaks (in the surge, sway, roll and pitch  directions) are caused by the relative motion of two floating bodies during the forced motions (they don’t actually exist because of the constraints), what should be done with it?

Cheers,
Tianyuan

**Comment by dforbush2** [comment]
Hi All,
[WAMITHelp_public.pdf](https://github.com/WEC-Sim/WEC-Sim/files/14364974/WAMITHelp_public.pdf)

I'm attaching a pdf of a powerpoint that shows our current understanding here. The quick summary is that the way we had meshed RM3 was ill-advised but "worked" in WAMIT v7.2. But in the v7.4 and later updates, the previous meshing approach results in these peaks that are undesirable for WEC-Sim, but these later WAMIT versions are more robust implementations of the solver so these results are apparently improved in this respect. I wish I had a clean, final, answer here but I think it will be a work in progress for some time. 

Best practices summary:
1) do not include moonpools unless they are a necessary, physical aspect of your device. There will be associated resonances related to its draft and radius that, without applying damping in the BEM model, will look like singularities. In the RM3, there isn't actually a moonpool, just a mated spar and float. In the improved mesh, the buoy is modeled as a solid.

2) Do not have interfering or overlapping mesh panels. In our example mesh the mated surface between the spar and the float interfered and this resulted in a poorly posed problem from the perspective of BEM.

3) Check with and without IRR: this appears to be acting counter-intuitively. 

4) increase mesh resolution. For the RM3 example, this was 2-3 times what is presented in the example files.

5) Use the badBEM_fix.m (on DEV Branch) cleanup function to remove things that look like singularities

**Comment by TianyuanWangi** [comment]
Hi [dforbush2](https://github.com/dforbush2) @dforbush2 
Many thanks for your detailed discussions. A question that may sound silly: How to create circular mesh faces in Rhinoceros like this? Could you please give me some references?
![image](https://github.com/WEC-Sim/WEC-Sim/assets/106663159/874887a0-8115-46c2-af26-699b8bb69a7d)

Best regards,
Tianyuan

**Comment by TianyuanWangi** [comment]
Hi [dforbush2](https://github.com/dforbush2) @dforbush2

I tested the code you shared. A powerful tool! It seems to be developed based on an interpolation method, and is effective in eliminating sparsely distributed singularities within a frequency range.

I encountered bugs and made slight modifications. For example, Line 161:
```
for kkk=1:length(BLocs)
    if BLocs(kkk) > 2 & BLocs(kkk) < length(hydro.w)-2
        BRep = interp1([hydro.w(BLocs(kkk)-2);hydro.w(BLocs(kkk)-1);hydro.w(BLocs(kkk)+1);hydro.w(BLocs(kkk)+2)],[hydro.B(k,kk,BLocs(kkk)-2);hydro.B(k,kk,BLocs(kkk)-1);hydro.B(k,kk,BLocs(kkk)+1);hydro.B(k,kk,BLocs(kkk)+2)],hydro.w(BLocs(kkk)),'linear');
        hydro.B(k,kk,BLocs(kkk))=BRep;
    end
end
```
If a singularity occurs at the boundary of the studied frequency range, no interpolation is performed.

**Comment by kmruehl** [comment]
@dforbush2 and @salhus migrate to discussion board
